File: packages/core/src/output-dir.ts
import fs from 'node:fs'
import path from 'node:path'

export interface ResolveAdapterOutputDirInput {
  /**
   * Honored as-is if set — used by adapters that expose a user-facing
   * `outputDir` option (e.g. WDIO). Skips all other resolution steps.
   */
  userConfiguredDir?: string
  /**
   * Absolute path to the current test file. When known, the video / trace
   * lands in the same folder as the spec the user just ran. This is the
   * preferred location across adapters.
   */
  testFilePath?: string
  /**
   * Absolute path to the resolved framework config file (wdio.conf.ts,
   * nightwatch.conf.cjs, etc.). Used as a fallback when the test file
   * isn't known.
   */
  configPath?: string
  /** Last-resort fallback. Defaults to `process.cwd()`. */
  fallbackDir?: string
}

const NODE_MODULES_SEGMENT = `${path.sep}node_modules${path.sep}`

function isWritable(dir: string): boolean {
  try {
    fs.accessSync(dir, fs.constants.W_OK)
    return true
  } catch {
    return false
  }
}

/**
 * Resolve the directory where an adapter should write output files
 * (screencast .webm, trace JSON, etc.).
 *
 * Priority:
 *   1. `userConfiguredDir` — explicit opt-in, honored as-is.
 *   2. `dirname(testFilePath)` — same folder as the spec that just ran.
 *   3. `dirname(configPath)` — fallback to the framework config dir.
 *   4. `fallbackDir` (default `process.cwd()`).
 *
 * Any candidate inside a `node_modules/` segment is skipped — this can
 * happen in symlinked workspaces where the test file resolves through a
 * linked dependency. Each candidate must also be writable; non-writable
 * dirs fall through to the next.
 *
 * Shared by all three adapters (service / nightwatch-devtools /
 * selenium-devtools) so the output location stays consistent regardless
 * of where the user invoked the runner from. See CLAUDE.md §2.2.
 */
export function resolveAdapterOutputDir(
  input: ResolveAdapterOutputDirInput = {}
): string {
  const fallback = input.fallbackDir ?? process.cwd()
  // userConfiguredDir bypasses the node_modules and writability filters
  // because the user opted into it explicitly — surprising overrides are
  // worse than failing loudly here.
  if (input.userConfiguredDir) {
    return input.userConfiguredDir
  }
  const candidates: string[] = []
  if (input.testFilePath) {
    candidates.push(path.dirname(input.testFilePath))
  }
  if (input.configPath) {
    candidates.push(path.dirname(input.configPath))
  }
  candidates.push(fallback)
  for (const dir of candidates) {
    if (!dir) {
      continue
    }
    if (dir.includes(NODE_MODULES_SEGMENT)) {
      continue
    }
    if (isWritable(dir)) {
      return dir
    }
  }
  return fallback
}

File: packages/selenium-devtools-py/src/selenium_devtools/output_dir.py
"""Where the adapter writes run output. Mirrors ``core/output-dir.ts`` so all four
adapters group artifacts under the same ``test-results/`` subfolder, wherever the
base directory resolves to.

Base directory, by priority:

    1. ``user_configured_dir`` — explicit opt-in, honored as-is.
    2. ``dirname(test_file_path)`` — the folder holding the test that just ran.
    3. ``fallback_dir`` (default: cwd).

A candidate inside an installed-package directory is skipped: a test file
resolved through site-packages must not receive artifacts. That is the Python
counterpart of core skipping ``node_modules``. Each candidate must also be
writable, or resolution falls through to the next.
"""

from __future__ import annotations

import os
from typing import List, Optional

#: All run output is grouped under this subfolder, matching the JS adapters.
OUTPUT_SUBDIR = "test-results"

#: Installed-package segments: the analogue of core's ``node_modules`` filter.
_INSTALLED_SEGMENTS = (
    f"{os.sep}site-packages{os.sep}",
    f"{os.sep}dist-packages{os.sep}",
)


def _is_writable(directory: str) -> bool:
    return os.path.isdir(directory) and os.access(directory, os.W_OK)


def _resolve_base_dir(
    user_configured_dir: Optional[str],
    test_file_path: Optional[str],
    fallback_dir: Optional[str],
) -> str:
    fallback = fallback_dir or os.getcwd()
    # An explicit directory bypasses the installed-package and writability
    # filters: the user opted into it, and a surprising override is worse than
    # failing where they pointed us.
    if user_configured_dir:
        return user_configured_dir
    candidates: List[str] = []
    if test_file_path:
        candidates.append(os.path.dirname(os.path.abspath(test_file_path)))
    candidates.append(fallback)
    for directory in candidates:
        if not directory:
            continue
        if any(seg in f"{directory}{os.sep}" for seg in _INSTALLED_SEGMENTS):
            continue
        if _is_writable(directory):
            return directory
    return fallback


def resolve_adapter_output_dir(
    *,
    user_configured_dir: Optional[str] = None,
    test_file_path: Optional[str] = None,
    fallback_dir: Optional[str] = None,
) -> str:
    """Absolute path of the ``test-results`` directory to write into. Does not
    create it; ``ensure_output_dir`` does that at write time."""
    base = _resolve_base_dir(user_configured_dir, test_file_path, fallback_dir)
    return os.path.join(os.path.abspath(base), OUTPUT_SUBDIR)


def ensure_output_dir(directory: str) -> Optional[str]:
    """Create ``directory`` if needed. Returns it, or None when it cannot be
    created, so a caller can fall back rather than crash a user's test run."""
    try:
        os.makedirs(directory, exist_ok=True)
    except OSError:
        return None
    return directory if _is_writable(directory) else None
